liquidhaskell-boot/src/Language/Haskell/Liquid/GHC/Misc.hs

```diff
@@ -1007,3 +1007,21 @@ prependGHCRealQual = varQual_RDR gHC_INTERNAL_REAL
 
 isFromGHCReal :: NamedThing a => a -> Bool
 isFromGHCReal x = Ghc.nameModule (Ghc.getName x) == gHC_INTERNAL_REAL
+
+-- Collect all free variables in a `CoreExpr`. The one in GHC libs is too restrictive, so
+-- we implement it ourselves here. Here, we only ingore the variables that are bound by lambda
+-- abstractions and nothing else.
+collectAllFreeVars :: CoreExpr -> S.HashSet Var
+collectAllFreeVars = go S.empty
+  where
+    go vars (Let (NonRec b rhs) e) = go (go (S.insert b vars) rhs) e
+    go vars (Let (Rec binds) e)    = foldr (\(b, rhs) vs -> go (S.insert b vs) rhs) (go vars e) binds
```

Comment:
I find odd that the variable names are collected when the bindings are found. If they were collected only when used, I think the function would be equally useful with less work.

Another odd thing is that bindings are only deleted if they appear as lambdas instead of `Let`s. Both bindings should be treated the same to claim that only free variables are returned.

Lastly, you might want to use `bindersOf` to handle the `Let` case.

Reply:
It's actually very possible that there is a function from ghc that does exactly what I want, but I could not find it so far

Reply:
Btw what do you mean by collected only when used?

Reply:
> Btw what do you mean by collected only when used?

`collectAllVars (let {a = 1; b = 2} in b) == [a, b]` where it would suffice to return only `[b]` because `a` is never used.